### DISCUSSION TITLE
I can't add className using element.classList.add(<className>)

### DISCUSSION THREAD

**Discussion by namnguyenthanhwork** [facebook/docusaurus#9104] [Q&A] [answered]
<img width="626" alt="image" src="https://github.com/facebook/docusaurus/assets/78526193/bb1f3e49-c77e-44e8-bd32-a177d7f070de">
<img width="732" alt="image" src="https://github.com/facebook/docusaurus/assets/78526193/6638f539-4078-4fcf-8de7-99f9b0e18e1c">

I have a component <LoaderPage/>, I want to user go to homepage or navigate to homepage will show loader a few second and disappear). It ok until LoaderPage component can't add className 'loaded' when <LoaderPage/> rendered. 
I want add class 'loaded' when <LoaderPage/> render. 
How to fix it. Thanks everyone !!!

**Comment by slorber** [comment]
You are not supposed to trigger side-effects in the render phase of React: this is against the very principles of React.

Using an effect is much more likely the solution, but I can't tell because you don't say why you want to show this loader in the first place.

**Reply by namnguyenthanhwork** [reply]
I want that when users visit the homepage, there will be an animation loading for a moment and then disappearing. I have css for this animation loader, and I need to add the class "loaded" when users visit the homepage so that this animation disappears. 
Do you have any ideas? 
I'm trying to use useEffect, but it doesn't work. No class is added to the body every time I visit the homepage.
Thanks @slorber !

**Reply by slorber** [reply, marked as answer]
Docusaurus is a static generator, so you first need to decide: should your loading class be in the static markup or not? If you want the loader to be displayed as soon as possible on first page load, it should be in the markup, and you can't add this using a `useEffect` because useEffect runs later and the loader won't be displayed immediately, leading to a weird UX.

---

Then why does it have to be applied to the body exactly? 

Why not any other dom element like this:

```tsx
const [loading, setLoading] = useState(true)

useTimeout(() => {
    setLoading(false);
  }, 1000);

<div className={loading ? "loading" : undefined}/>
```

https://www.joshwcomeau.com/snippets/react-hooks/use-timeout/

---

If you really need it to be added to the body in the static html markup, you have those options:
- customize the SSR template (applies to all pages): https://docusaurus.io/docs/next/api/docusaurus-config#ssrTemplate
- use `<Head><body className="loading"></Head>` (can be applied to a single page), see https://docusaurus.io/docs/docusaurus-core#head

---

Why does it have to be "loading"? It would be much simpler to add a "loaded" class after React hydration:

```tsx
useTimeout(() => {
    document.body.classList.add("loaded")
  }, 1000);
```

Instead of using CSS: `body.loading`
You could use a different CSS selector to achieve the same: `html.homepage > body:not(.loaded)`

**Reply by namnguyenthanhwork** [reply]
Thanks for your help; it's useful for me.
